plugins/languages/cmake-tools.nix

```
@@ -0,0 +1,391 @@
+{
+  lib,
+  helpers,
+  config,
+  pkgs,
+  ...
+}:
+helpers.neovim-plugin.mkNeovimPlugin config {
+  name = "cmake-tools";
+  originalName = "cmake-tools.nvim";
+  defaultPackage = pkgs.vimPlugins.cmake-tools-nvim;
+
+  maintainers = [helpers.maintainers.NathanFelber];
+
+  extraPlugins = with pkgs.vimPlugins; [
+    plenary-nvim
+  ];
+
+  settingsOptions = {
+    cmake_command = helpers.defaultNullOpts.mkStr "cmake" ''
+      This is used to specify cmake command path.
+    '';
+
+    ctest_command = helpers.defaultNullOpts.mkStr "ctest" ''
+      This is used to specify ctest command path.
+    '';
+
+    cmake_regenerate_on_save = helpers.defaultNullOpts.mkBool true ''
+      Auto generate when save CMakeLists.txt.
+    '';
+
+    cmake_generate_options =
+      helpers.defaultNullOpts.mkAttributeSet ''
+        { "-DCMAKE_EXPORT_COMPILE_COMMANDS" = 1; }
+      ''
+      ''
+        This will be passed when invoke `CMakeGenerate`.
+      '';
+
+    cmake_build_options = helpers.defaultNullOpts.mkAttributeSet "{}" ''
+      This will be passed when invoke `CMakeBuild`.
+    '';
+
+    cmake_build_directory = helpers.defaultNullOpts.mkStr "out/\${variant:buildType}" ''
+      This is used to specify generate directory for cmake, allows macro expansion, relative to vim.loop.cwd().
+    '';
+
+    cmake_soft_link_compile_commands = helpers.defaultNullOpts.mkBool true ''
+      This will automatically make a soft link from compile commands file to project root dir.
+    '';
+
+    cmake_compile_commands_from_lsp = helpers.defaultNullOpts.mkBool false ''
+      This will automatically set compile commands file location using lsp, to use it, please set `cmake_soft_link_compile_commands` to false.
+    '';
+
+    cmake_kits_path = helpers.defaultNullOpts.mkStr null ''
+      This is used to specify global cmake kits path, see CMakeKits for detailed usage.
+    '';
+
+    cmake_variants_message =
+      helpers.mkCompositeOption ''
+        Settings for cmake variants messages.
+      ''
+      {
+        short =
+          helpers.defaultNullOpts.mkAttributeSet ''
+            { show = true; }
+          ''
+          ''
+            Whether to show short message.
+          '';
+
+        long =
+          helpers.defaultNullOpts.mkAttributeSet ''
+            { show = true; max_length = 40; }
+          ''
+          ''
+            Whether to show long message.
+          '';
+      };
+
+    cmake_dap_configuration =
+      helpers.defaultNullOpts.mkAttributeSet ''
+        {
+          name = "cpp";
+          type = "codelldb";
+          request = "launch";
+          stopOnEntry = false;
+          runInTerminal = true;
+          console = "integratedTerminal";
+        }
+      ''
+      ''
+        Dap configuration for cmake.
+      '';
```

```suggestion
    cmake_dap_configuration =
      helpers.defaultNullOpts.mkAttrsOf lib.types.anything {
          name = "cpp";
          type = "codelldb";
          request = "launch";
          stopOnEntry = false;
          runInTerminal = true;
          console = "integratedTerminal";
        } ''
        Dap configuration for cmake.
      '';
```

- We no longer quote plugin-defaults
- `attrsOf anything` is preferred over `attrs`
- If `name`, `type`, `request`, etc are well known options, you could make them actual options instead. Or define a submodule type to use here.

---

In this case, `name`, `type` and `request` are well known required options while the others are optional and depend on the debug adapter being used (e.g. `codelldb`). It would be nice to package the first three but also let the user add arbitrary attributes, and I'm not sure how to properly do that. mkCompositeOption doesn't seem to take a plugin default, and mkAttrsOf anything doesn't seem to take child options. Any idea?

---

> It would be nice to package the first three but also let the user add arbitrary attributes, and I'm not sure how to properly do that

The `settings` option has a "freeform" type (`attrsOf anything`), so this is the default behaviour, unless we add additional restrictions via option types.

If that's what you want here then all you need to do is:

```nix
cmake_dap_configuration = {
  name = /*mkOption*/;
  type = /*mkOption*/;
  request = /*mkOption*/;
};
```

If this needs to be done in a re-usable way, then there's a few ways to do this. The simplest would be to define the option attrs in a let block and re-use it that way.

With this approach you could also define "extra" options using the `commonOpts // {/* extraOpts */}` pattern (if you wanted to).

---

It's also possible to do this kinda thing with submodules, however `mkCompositeOption` is probably not the best way to define a submodule option, because it doesn't set a freeform type and AFAIK submodules don't inherit the parent's freeform type.

A submodule option _would_ allow you to define things like `example`, `description` and `pluginDefault` though.

I'd suggest ignoring submodules for now, focus on getting something simple working and then revisit the idea later if you're not satisfied with the end result.

---

I see, thank you for the clarification. Then I'll go the simplest route and only declare these three options as you suggest.